File: tests/testshape.cc
// -*- LSST-C++ -*-

/* 
 * LSST Data Management System
 * Copyright 2008, 2009, 2010 LSST Corporation.
 * 
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the LSST License Statement and 
 * the GNU General Public License along with this program.  If not, 
 * see <http://www.lsstcorp.org/LegalNotices/>.
 */
 
//
// test a perfect Gaussian PSF and measure aperture photometry at different radii
//
#include <iostream>
#include <sstream>
#include <string>
#include <limits>
#include <cmath>
#include "lsst/pex/policy/Policy.h"
#include "lsst/afw.h"
#include "lsst/afw/detection/Psf.h"
#include "lsst/afw/image/ImageAlgorithm.h"
#include "lsst/meas/algorithms/Measure.h"
#include "lsst/afw/math/Integrate.h"
#include "lsst/afw/coord/Coord.h"
#include "lsst/meas/extensions/shapeHSM/HsmShapeControl.h"
#include "lsst/meas/algorithms/SingleGaussianPsf.h"

#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_MODULE HsmShape

#include "boost/test/unit_test.hpp"
#include "boost/test/floating_point_comparison.hpp"

using namespace std;
namespace pexPolicy = lsst::pex::policy;
namespace measAlg = lsst::meas::algorithms;
namespace afwDetection = lsst::afw::detection;
namespace afwImage = lsst::afw::image;
namespace afwMath = lsst::afw::math;
namespace afwTable = lsst::afw::table;
namespace afwGeom = lsst::afw::geom;
namespace afwCoord = lsst::afw::coord;
namespace hsm = lsst::meas::extensions::shapeHSM;

typedef afwImage::Exposure<float, short unsigned int, float> ExposureT;
typedef ExposureT::MaskedImageT MImage;

BOOST_AUTO_TEST_CASE(HsmMoments) {

    float bkgd = 100.0;
    int xwidth = 100;
    int ywidth = 100;
    MImage mimg(afwGeom::Extent2I(xwidth, ywidth));
    *mimg.getImage() = bkgd;

    int x = 30;
    int y = 40;
    float I0 = 1000.0;

    // make a masked image and an exposure
    (*mimg.getImage())(x,y) = I0 + bkgd;
    *mimg.getMask() = 0;
    *mimg.getVariance() = sqrt(I0);
    ExposureT::Ptr exposure(new ExposureT(mimg));

    int kwid = 35;
    float sigma = 1.0;
    afwDetection::Psf::Ptr psf(new measAlg::SingleGaussianPsf(kwid, kwid, sigma));
    exposure->setPsf(psf);
    exposure->setWcs(
        afwImage::makeWcs(
            *afwCoord::makeCoord(afwCoord::ICRS, 0.0 * afwGeom::degrees, 0.0 * afwGeom::degrees),
            afwGeom::Point2D(1.0, 1.0),
            0.2/3600.0, 0.0, 0.0, 0.2/3600.0
        )
    );
    
    // Add a Gaussian to the image
    float sigma_xx = std::pow(1.5, 2);
    float sigma_xy = 0;
    float sigma_yy = std::pow(2.5, 2);
    int ksize = 15;


    // compute the known values of the moments
    float phi = 0.0;  // rotate +ve this far (in radians)
    float c = cos(phi);
    float s = sin(phi);

    float sum, sumxx, sumxy, sumyy;
    sum = sumxx = sumxy = sumyy = 0.0;

    for (int dx = -(ksize/2); dx < ksize/2 + 1; dx++) {
        for (int dy = -(ksize/2); dy < ksize/2 + 1; dy++) {
            float u = c*dx + s*dy;
            float v = s*dx + c*dy;
            float I = I0*exp(-0.5*(u*u/sigma_xx + v*v/sigma_yy));
            MImage::iterator ptr = mimg.at(x+dx, y+dy);
            ptr.image() = bkgd + I;
            sum += I;
            sumxx += I*dx*dx;
            sumxy += I*dx*dy;
            sumyy += I*dy*dy;
        }
    }
    sumxx /= sum;
    sumxy /= sum;
    sumyy /= sum;
        
    printf("%g %g %g\n", sumxx, sumxy, sumyy);


    // measure the shape with our algorithms
    std::vector<PTR(measAlg::AlgorithmControl)> algList;
    algList.push_back(boost::make_shared<hsm::HsmShapeBjControl>());
    algList.push_back(boost::make_shared<hsm::HsmShapeLinearControl>());
    algList.push_back(boost::make_shared<hsm::HsmShapeKsbControl>());
    algList.push_back(boost::make_shared<hsm::HsmShapeRegaussControl>());
    algList.push_back(boost::make_shared<hsm::HsmShapeShapeletControl>());

    PTR(MImage::Image) img = exposure->getMaskedImage().getImage();
    *img -= bkgd;

    for (
        std::vector<PTR(measAlg::AlgorithmControl)>::iterator it = algList.begin();
        it != algList.end();
        ++it
    ) {
        
        afwTable::Schema schema = afwTable::SourceTable::makeMinimalSchema();
        measAlg::MeasureSources measureShape =
            measAlg::MeasureSourcesBuilder().addAlgorithm(**it).build(schema);
        PTR(afwTable::SourceTable) table = afwTable::SourceTable::make(schema);

        afwGeom::Point2D center(x, y);
        PTR(afwDetection::Footprint) foot = boost::make_shared<afwDetection::Footprint>(exposure->getBBox());
        PTR(afwTable::SourceRecord) source = table->makeRecord();
        source->setFootprint(foot);

        measureShape.apply(*source, *exposure, center);

        afwTable::SubSchema s = schema[(**it).name];
        afwGeom::Point2D p = source->get< afwTable::Point<double> >(s["centroid"]);
        afwGeom::ellipses::Quadrupole q = source->get< afwTable::Moments<double> >(s["moments"]);

        // compare to known values
        float Ixx = q.getIxx();
        float Iyy = q.getIyy();
        float Ixy = q.getIxy();
        float A2 = 0.5*(Ixx + Iyy) + sqrt( pow(0.5*(Ixx - Iyy), 2) + pow(Ixy, 2) );
        float B2 = 0.5*(Ixx + Iyy) - sqrt( pow(0.5*(Ixx - Iyy), 2) + pow(Ixy, 2) );

        printf("Algorithm: %s\n", (**it).name.c_str());
        printf("x:     %d %.2f\n", x, p.getX());
        printf("y:     %d %.2f\n", y, p.getY());
        printf("I_xx:  %.5f %.5f\n", sigma_xx, Ixx);
        printf("I_xy:  %.5f %.5f\n", sigma_xy, Ixy);
        printf("I_yy:  %.5f %.5f\n", sigma_yy, Iyy);
        printf("A2, B2 = %.5f, %.5f\n", A2, B2);      
    
        // 1/100 pixel limit for x,y
        BOOST_CHECK(fabs(static_cast<double>(x) - p.getX()) < 1e-2);
        BOOST_CHECK(fabs(static_cast<double>(y) - p.getY()) < 1e-2);

        // higher limit for ixx/xy/yy
        BOOST_CHECK(fabs(Ixx - sigma_xx) < 0.02*(1.0+sigma_xx));
        BOOST_CHECK(fabs(Ixy - sigma_xy) < 0.02*(1.0+sigma_xy));
        BOOST_CHECK(fabs(Iyy - sigma_yy) < 0.03*(1.0+sigma_yy));
        
    }
}
    

File: tests/testshape.py
#!/usr/bin/env python
# 
# LSST Data Management System
# Copyright 2008, 2009, 2010 LSST Corporation.
# 
# This product includes software developed by the
# LSST Project (http://www.lsst.org/).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the LSST License Statement and 
# the GNU General Public License along with this program.  If not, 
# see <http://www.lsstcorp.org/LegalNotices/>.
#

import re, os, sys
import glob
import math
import unittest

import lsst.pex.policy as pexPolicy
import lsst.pex.exceptions as pexExceptions
import lsst.afw.image as afwImage
import lsst.afw.math as afwMath
import lsst.afw.geom as afwGeom
import lsst.meas.algorithms as algorithms
import lsst.utils.tests as utilsTests
import lsst.afw.detection as afwDetection
import lsst.afw.table as afwTable

import lsst.afw.display.ds9 as ds9

import lsst.meas.extensions.shapeHSM as hsm

try:
    type(verbose)
except NameError:
    verbose = 0
    display = False

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

class HsmMomentTestCase(unittest.TestCase):
    """A test case for the HSM shape moments (un-psf-corrected)."""

    def setUp(self):
        
        self.bkgd = 100.0
        self.x, self.y = 30, 40
        self.I0 = 1000.0

        # make a masked image and an exposure
        im = afwImage.ImageF(afwGeom.Extent2I(100, 100))
        im.set(self.bkgd)
        im.set(self.x, self.y, self.I0 + self.bkgd)
        msk = afwImage.MaskU(im.getDimensions()); msk.set(0)
        var = afwImage.ImageF(im.getDimensions()); var.set(math.sqrt(self.I0))
        self.mimg = afwImage.MaskedImageF(im, msk, var)
        del im; del msk; del var
        self.exposure = afwImage.makeExposure(self.mimg)
        
        kwid = 35
        sigma = 2.5
        self.psf = algorithms.SingleGaussianPsf(kwid, kwid, sigma)

        self.exposure.setPsf(self.psf)
        
        # Add a Gaussian to the image
        self.sigma_xx, self.sigma_xy, self.sigma_yy, self.ksize = math.pow(1.5, 2), 0, math.pow(2.5, 2), 15

        phi = math.radians(0.0)      # rotate +ve this far
        c, s = math.cos(phi), math.sin(phi)

        self.sum, self.sumxx, self.sumxy, self.sumyy = 4*[0.0]
        for dx in range(-(self.ksize/2), self.ksize/2 + 1):
            for dy in range(-(self.ksize/2), self.ksize/2 + 1):
                u, v = c*dx + s*dy,  s*dx - c*dy
                I = self.I0*math.exp(-0.5*(u*u/self.sigma_xx + v*v/self.sigma_yy))
                self.mimg.getImage().set(self.x + dx, self.y + dy, self.bkgd + I)

                self.sum += I
                self.sumxx += I*dx*dx
                self.sumxy += I*dx*dy
                self.sumyy += I*dy*dy

        self.sumxx /= self.sum
        self.sumxy /= self.sum
        self.sumyy /= self.sum


        self.policy = pexPolicy.Policy(pexPolicy.PolicyString(
                """#<?cfg paf policy?>
            HSM_SHAPELET.max_order_psf: 8
            HSM_SHAPELET.max_order_gal: 8
            """
                ))
        
        print "%g %g %g" % (self.sumxx, self.sumxy, self.sumyy)


        
    def tearDown(self):
        del self.mimg
        del self.exposure
        del self.psf
        del self.policy


        
    def testHsmMoments(self):
        """Test that the un-psf-corrected moments are correct"""

        # measure shape with our algorithm
        algorithmNames = ["shape.hsm.bj", "shape.hsm.linear", "shape.hsm.ksb",
                          "shape.hsm.regauss", "shape.hsm.shapelet"]

        mim = self.exposure.getMaskedImage()
        im  = mim.getImage()
        im -= self.bkgd
        
        for algName in algorithmNames:
            schema = afwTable.SourceTable.makeMinimalSchema()
            msConfig = algorithms.SourceMeasurementConfig()
            msConfig.algorithms.names = [algName]
            shapeFinder = msConfig.makeMeasureSources(schema)

            if display:
                ds9.mtv(self.mimg)

            center = afwGeom.Point2D(self.x, self.y)
            table = afwTable.SourceTable.make(schema)
            source = table.makeRecord()
            source.setFootprint(afwDetection.Footprint(self.exposure.getBBox()))

            shapeFinder.apply(source, self.exposure, center)

            p = source.get(algName + ".centroid")
            s = source.get(algName + ".moments")

            # see how we did
            Ixx, Iyy, Ixy = s.getIxx(), s.getIyy(), s.getIxy()
            A2 = 0.5*(Ixx + Iyy) + math.sqrt( (0.5*(Ixx - Iyy))**2 + Ixy**2 )
            B2 = 0.5*(Ixx + Iyy) - math.sqrt( (0.5*(Ixx - Iyy))**2 + Ixy**2 )

            print "I_xx:  %.5f %.5f" % (Ixx, self.sigma_xx)
            print "I_xy:  %.5f %.5f" % (Ixy, self.sigma_xy)
            print "I_yy:  %.5f %.5f" % (Iyy, self.sigma_yy)
            print "A2, B2 = %.5f, %.5f" % (A2, B2)         

            # 1/100 pixel limit
            self.assertTrue(abs(self.x - p.getX()) < 1e-2, "%g v. %g" % (self.x, p.getX()))
            self.assertTrue(abs(self.y - p.getY()) < 1e-2, "%g v. %g" % (self.y, p.getY()))

            # need higher limit for these (they're just placeholder shapes computed, after all)
            self.assertTrue(abs(s.getIxx() - self.sigma_xx) < 2e-2*(1 + self.sigma_xx),
                            "%g v. %g" % (self.sigma_xx, s.getIxx()))
            self.assertTrue(abs(s.getIxy() - self.sigma_xy) < 2e-2*(1 + self.sigma_xy),
                            "%g v. %g" % (self.sigma_xy, s.getIxy()))
            self.assertTrue(abs(s.getIyy() - self.sigma_yy) < 3e-2*(1 + self.sigma_yy),
                            "%g v. %g" % (self.sigma_yy, s.getIyy()))


        
        
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

def suite():
    """Returns a suite containing all the test cases in this module."""
    utilsTests.init()

    suites = []
    suites += unittest.makeSuite(HsmMomentTestCase)
    suites += unittest.makeSuite(utilsTests.MemoryTestCase)

    return unittest.TestSuite(suites)

def run(exit = False):
    """Run the tests"""
    utilsTests.run(suite(), exit)
 
if __name__ == "__main__":
    run(True)
